File: 2-NORMAL/JavaScript/2-merge-sort-implementation/proposed-solution/main.js

/*
Challenge: Develop a function that implements the Merge Sort algorithm to sort an array.
Evaluate the efficiency of your implementation and consider cases with large arrays.
*/

// main.js - Example usage of MergeSort

import MergeSort from './mergeSort.js';

function main() {
    const sampleArray = [38, 27, 43, 3, 9, 82, 10];
    console.log("Original array:", sampleArray);
    const sortedArray = MergeSort.mergeSort(sampleArray);
    console.log("Sorted array:", sortedArray);
}

main();

File: 2-NORMAL/Python/2-merge-sort-implementation/proposed-solution/main.py
# Challenge: Develop a function that implements the Merge Sort algorithm to sort an array.
# Evaluate the efficiency of your implementation and consider cases with large arrays.

def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result

# Example usage
if __name__ == "__main__":
    sample_array = [38, 27, 43, 3, 9, 82, 10]
    print("Original array:", sample_array)
    sorted_array = merge_sort(sample_array)
    print("Sorted array:", sorted_array)
